apps/aecore/src/aec_conductor.erl

```diff
@@ -544,48 +544,66 @@ start_mining(#state{block_candidate = #candidate{top_hash = OldHash},
     %% Candidate generated with stale top hash.
     %% Regenerate the candidate.
     create_block_candidate(State#state{block_candidate = undefined});
-start_mining(#state{} = State) ->
+start_mining(#state{block_candidate = Candidate} = State) ->
     epoch_mining:info("Starting mining"),
-    #candidate{block = Block,
-               nonce = Nonce} = State#state.block_candidate,
-    Info = [{top_block_hash, State#state.seen_top_block_hash}],
+    BlockBin = Candidate#candidate.bin,
+    Nonce    = Candidate#candidate.nonce,
+    Target   = aec_blocks:target(Candidate#candidate.block),
+    Info     = [{top_block_hash, State#state.seen_top_block_hash}],
     aec_events:publish(start_mining, Info),
-    Fun = fun() -> {aec_mining:mine(Block, Nonce),
-                    State#state.block_candidate}
+    Fun = fun() ->
+                  {aec_mining:mine(BlockBin, Target, Nonce)
+                  , BlockBin}
           end,
     dispatch_worker(mining, Fun, State).
 
-handle_mining_reply({{ok, Block},_Candidate}, State) ->
-    aec_metrics:try_update([ae,epoch,aecore,mining,blocks_mined], 1),
-    State1 = State#state{block_candidate = undefined},
-    case handle_mined_block(Block, State1) of
-        {ok, State2} ->
-            State2;
-        {{error, Reason}, State2} ->
-            epoch_mining:error("Block insertion failed: ~p.", [Reason]),
-            start_mining(State2)
+handle_mining_reply(_Reply, #state{block_candidate = undefined} = State) ->
+    %% Something invalidated the block candidate already.
+    start_mining(State);
+handle_mining_reply({{ok, {Nonce, Evd}}, BlockBin}, #state{} = State) ->
+    Candidate = State#state.block_candidate,
+    %% Check that the solution is for this block
+    case BlockBin =:= Candidate#candidate.bin of
+        true ->
+            aec_metrics:try_update([ae,epoch,aecore,mining,blocks_mined], 1),
+            State1 = State#state{block_candidate = undefined},
+            Block = aec_blocks:set_pow(Candidate#candidate.block, Nonce, Evd),
+            case handle_mined_block(Block, State1) of
+                {ok, State2} ->
+                    State2;
+                {{error, Reason}, State2} ->
+                    epoch_mining:error("Block insertion failed: ~p.", [Reason]),
+                    start_mining(State2)
+            end;
+        _Other ->
+            %% This mining effort was for an earlier block candidate.
+            epoch_mining:error("Found solution for old block", []),
```

Comment:
As the pid here is the one of the conductor - not the worker - and this case shall not happen often in my understanding, maybe operator would like the nonce logged here for correlating event that made this happen? (`Candidate#candidate.bin` may be too much info.)

Reply:
This is the case when a new candidate has been built because of some trigger (most likely, the top changed), but the old mining had already sent the answer. I'm not even sure that this can happen, since it should show up as an "unsolicited worker answer", but it is a bit of a safe guard.